django/contrib/postgres/search.py

```diff
@@ -178,17 +178,22 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize
         resolved = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
         if self.config:
             resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        if hasattr(self.value, 'resolve_expression'):
+            resolved.value = self.value.resolve_expression(query, allow_joins, reuse, summarize, for_save)
         return resolved
```

Comment:
This makes me wonder why `SearchQuery` is a `Value` in the first place given it needs to resolve `config` and `value` now.

Reply:
To be honest I'm a bit lost between `Value`, `Expression`, `Func`, ...
What do you think would be a better base class for this?

Reply:
If look like could subclass `Func` and avoid any `resolve_expression` or `as_sql` overrides.

```python
class SearchQuery(SearchQueryCombinable, Func):
    output_field = SearchQueryField()
    SEARCH_TYPES = {
        'plain': 'plainto_tsquery',
        'phrase': 'phraseto_tsquery',
        'raw': 'to_tsquery',
        'websearch': 'websearch_to_tsquery',
    }
    def __init__(value, output_field=None, *, config=None, invert=False, search_type='plain'):
        function = self.SEARCH_TYPES.get(search_type)
        if function is None:
            raise ValueError("Unknown search_type argument '%s'." % search_type)
        if not hasattr(value, 'resolve_expression'):
           value = Value(value)
        expressions = (value,)
        if config is not None:
            config = SearchConfig.from_parameter(config)
            expressions = (config,) + expressions
       super().__init__(*expressions, output_field=output_field, function=function)
        if invert:
            self.template = '!!(%s)' % Func.template
```

Reply:
Not knowing much about the internals, I think changing to `Func` makes sense.

I pushed a new commit that makes `SearchQuery` inherit from `Func` using your proposal but I had to make a few more changes to fix some failing `postgres_tests` tests.

There's still an issue with the tests for `SearchQuery.__repr__` (see commented out tests in the PR) but I don't think it's fixable other than by changing the expected output of the tests to the new one.

I'm also concerned about potential backwards compatibility issues. I'm not sure how much of the API of `SearchQuery` is considered public. It shouldn't be too hard to add deprecated properties for attributes like `config`, `search_type`, `invert`, ...
Do you think it's worth the effort?